### ISSUE TITLE
Not useful?

### ISSUE THREAD

**Issue by joelmccracken** [zimbatm/namespace.rb#1]
The readme mentions that it isnt clear why this is useful, but I think it would be very useful. 

Examples off the top of my head:
1. Multiple versions of libraries
2. Loading a library is not a global action, and thus there is less to worry about

**Comment by zimbatm** [comment]
Yes I agree. It's probably me who wasn't really sure if this was useful at the time of writing this.

Another benefit could be the ability to load a new version of the code without replacing the old version immediately.

The sandboxing is not good enough right now because class and module definitions don't inherit the namespace and this would probably cause confusion to the user. I haven't given it much thought but I think it would require to change some of the language.

**Comment by joelmccracken** [comment]
Can you give an example of what you're referring to?

**Comment by joelmccracken** [comment]
The general "globalness" of classes imo is one of the biggest problems facing Ruby today; this solves that problem.

**Comment by zimbatm** [comment]
I agree. It's just that there are some corner cases to handle before I consider this production ready. I would love to see more adoption on this.

For example the class names aren't particularly nice because they are defined in the context of an anonymous module. I don't think this can be fixed in MRI:

```
$ cd namespace.rb
$ irb -Ilib
irb(main):001:0> require 'namespace'
=> true
irb(main):002:0> Namespace.open('test/fixtures/b')
=> test::fixtures::b
irb(main):003:0> b = _
=> test::fixtures::b
irb(main):004:0> b::B
=> #<Module:0x007f974481a3e8>::B
```

Another issue is that the scope under a class definition doesn't inherit from it's parent scope. This means that things you've imported from another scope aren't accessible under the class. I think there's a bit more work required in the import/export scenarios.

``` ruby
x = "foo"
class X; puts x; end
#=> NameError: undefined local variable or method `x' for X:Class
```

If you want to come step up and take ownership, be my guest. I think it's a worthwhile project and would love to see more community adoption, it's just that I don't have the time at the moment.

**Comment by joelmccracken** [comment]
Thanks. I'd like to at least work on it or think about it. At the moment I
also don't have time, but I would like to get around to it sometime.

The parent ownership problem you mention is a Ruby limitation -- the class
keyword changes scope. However,

X = Class.new { puts x }

works fine. It IS uglier, though. There are a number of ways to tackle this
problem though, so its not _that_ bad.

Also, it seems like singleton method on the module might fix the ugly
naming problem. Not sure though.

I think this is very important for Ruby, and also would love to see it
developed.

**Comment by zimbatm** [comment]
Just repeating but you're very welcome if you want to.

**Comment by zimbatm** [comment]
Just saw this other interesting project that seem related and might interest you: https://github.com/soveran/cargo

**Comment by joelmccracken** [comment]
Nice, Thanks. FYI at the moment I have plans to hack on this, but I have to do a few other things first. I will certainly look at that project to see if I want to integrate anything from it.

**Comment by joelmccracken** [comment]
http://redshift.sourceforge.net/script/doc/index.html

^^ another similar-ish lib
